designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+include:
+- `validate_cmdstan` : determine whether the CmdStan installation works
+- `executable_works` : is there a working executable at the path
+  implied by the Stan file?
+- `compile_with_args` : attempt to compile a Stan program with
+  optional `make` arguments
+- `call_executable` : call executable with the given argument tree; on
+  success, return a `CmdStanOutput` instance.
+
+## Output
+
+Output of a successful `call_executable` call on a `CmdStanModel` will
+produce a `CmdStanOuput` instance, which encapsulates the context of
+the call. This includes:
+- the console output (exit status, stdout and stderr),
+- the argument tree provided to `call_executable`
+- the current working directory of the process at the time the call was made
+- the CmdStan installation from the parent `CmdStanModel`
+
+The objective is for `CmdStanOutput` to be a self-contained record
+which includes all pertinent information from a successful executable
+call. This structure can then be used to direct calls of
+`diagnose`/`stansummary`. Naturally, methods on said type will be
+present to expose the context to the user and perform utility
+functions (e.g. return a list of output file paths).
+
+## Arguments and options
+
+Stan provides several inference engines, each with a large number of
+options. CmdStan in turn handles this heterogeneity.
+
+To encapsulate the arguments passed at the command line (to a compiled
+executable), the proposal is a strongly-typed representation of this
+heterogeneity using a combination of sum types (Rust `enum`) and
+product types (Rust `struct`). By construction, this representation
+prevents the formation of inconsistent argument combinations -- the
+code simply won't compile. The resultant tree is an abstraction which
+enables the use of a single type (`CmdStanOutput`) to encapsulate a
+call to an executable.
+
+Unsurprisingly, the argument tree is a syntax tree for CmdStan
+command arguments. We translate to the very simple command line
+language, but leave open the possibility of translation to other
+languages.
+
+### Translation
+
+The (sloppy) productions for the command line language are:
+```text
+tree    -> terms
+terms   -> terms " " term | term
+term    -> pair | product | sum
+pair    -> key "=" value
+product -> type " " pairs
+sum     -> type "=" variant " " terms | type "=" variant
+pairs   -> pairs " " pair | pair
+
+key     -> A
+A       -> A alpha | beta
+alpha   -> letter | digit | "_"
+beta    -> letter
+letter  -> "A" | ... | "z"
+digit   -> "0" | ... | "9"
+
+value   -> number | path
+```
+Where the productions for `number` and `path` are left out for
+brevity.  The start symbol is `tree`. Generate the command line
+statement by folding the tree from left to right by generating the
+appropriate term from each node, building up a linear argument list.
+I sketched
+[this](https://github.com/andrewjradcliffe/cmdstan-translator/blob/main/translate.scm)
+out in Scheme, why I am not sure.
```

Comment:
I'm not sure this section is necessary, providing an example of use and/or a list of expected structs may be more useful